Reverse cs_main, cs_wallet lock order and reduce cs_main locking

ariard opened the pull request:
This change is intended to make the bitcoin node and its rpc, network and gui interfaces more responsive while the wallet is in use. Currently, because the node's `cs_main` mutex is always locked before the wallet's `cs_wallet` mutex (to prevent deadlocks), `cs_main` currently stays locked while the wallet does relatively slow things like creating and listing transactions.

Switching the lock order so `cs_main` is acquired after `cs_wallet` allows `cs_main` to be only locked intermittently while the wallet is doing slow operations, so the node is not blocked waiting for the wallet.

To review the present PR, most of getting right the move is ensuring any `LockAssertion` in `Chain::Lock` method is amended as a `LOCK(cs_main)`. And in final commit, check that any wallet code which was previously locking the chain is now calling a  method, enforcing the lock taking job. So far the only exception I found is `handleNotifications`, which should be corrected.

Commit:
[wallet] Move getHeight from Chain::Lock interface to simple Chain

Instead of calling getHeight, we rely on CWallet::m_last_block
processed_height where it's possible.

Commit:
[wallet] Move getBlockHeight from Chain::Lock interface to simple Chain

Add HaveChain to assert chain access for wallet-tool in LoadToWallet.

Commit:
[wallet] Move getBlockHash from Chain::Lock interface to simple Chain

Commit:
[wallet] Move methods from Chain::Lock interface to simple Chain

Remove findPruned and findFork, no more used after 17954.

ryanofsky commented:
Great work! I'd suggest changing the PR title to something like "Reduce wallet cs_main locking and reverse cs_main, cs_wallet lock order" to motivate it better and describe the change.

The current title "Remove Chain::Lock interface" and starting sentence "Follow-up in the set of Chain interface refactoring" make this sound mostly like a refactoring. But this is more of a locking change, and a change to make the wallet more asynchronous and event driven than a refactoring change.

practicalswift commented:
Concept ACK

Excellent work!

MarcoFalke commented:
Concept ACK. This might help IBD, because `cs_main` had to be acquired due to the lock order requirement. If it doesn't help IBD, the change will hopefully still speed up the `msghand` thread because the wallets take the main lock less.

jnewbery commented:
Big concept ACK! Thanks @ariard

promag commented:
Concept ACK, mother of god, not locking `cs_main`!

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -2274,13 +2252,13 @@ bool CWalletTx::InMempool() const
 {
     // Quick answer in most cases
-    if (!locked_chain.checkFinalTx(*tx)) {
+    if (!pwallet->chain().checkFinalTx(*tx)) {
```
Instead of adding an `interfaces::Chain::checkFinalTx` method, it might be better to call [`IsFinalTx`](https://github.com/bitcoin/bitcoin/blob/50591f6ec61b802cf4193cdbefcc85ad75716e8d/src/consensus/tx_verify.cpp#L17-L28) directly with the wallet's height and time, avoiding going through [`CheckFinalTx`](https://github.com/bitcoin/bitcoin/blob/50591f6ec61b802cf4193cdbefcc85ad75716e8d/src/validation.cpp#L189-L219).

This could perform better since it wouldn't require locking cs_main, and could make wallet calls return more internally consistent information when the wallet is catching up to the chain.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/16426#discussion_r342719832

> Instead of adding an `interfaces::Chain::checkFinalTx` method, it might be better to call [`IsFinalTx`](https://github.com/bitcoin/bitcoin/blob/50591f6ec61b802cf4193cdbefcc85ad75716e8d/src/consensus/tx_verify.cpp#L17-L28)

#17443 is a step in this direction (https://github.com/bitcoin/bitcoin/pull/17443#pullrequestreview-332692782)

ariard commented:
Finally rebased after merge of #15931, PR should be ready for review. 

Apart of 3efe38d which use the new `m_last_block_processed_height` to avoid querying the chainstate and introduce some modifications, other commits are pretty straight-forward. It's just taking `cs_main` inside the `Chain` implementation instead of using `Chain::lock`. Lock order is reversed only in last commit f057aed to avoid any deadlock issue.

If you still feel PR is hard to review, I can subsplit easily in another PR.

ryanofsky commented:
Approach ACK. Code changes here are very simple after #15931.

All the changes before the last commit f057aedf80d8bd6083c2227e42d4887be4c2933b seem straightforward and don't change behavior other than locking `cs_main` in new places recursively, so all the new locks are no-ops.

Only the last commit f057aedf80d8bd6083c2227e42d4887be4c2933b is the big scary change that removes `cs_main` locks all over the wallet, leaving us to hope that remaining locking is sufficient and that stretches of wallet code that used to run under `cs_main` aren't making assumptions about the tip not changing and other wallet threads not running.

---

I'll review this PR more closely this week. It might be nice to simplify the PR description now that #15931 is merged. I think the description just basically needs to say that `interfaces::Chain::Lock` is a wrapper around around `cs_main`, and that this PR changes wallet code to not lock and unlock `cs_main` directly anymore, and not keep `cs_main` locked while it locks `cs_wallet` and does other work. Instead, after this PR, wallet code only locks `cs_main` indirectly and intermittently when it needs to look up bits of chain information, and never holds onto `cs_main` while it does other work.

jnewbery commented:
Restating my concept ACK. I plan to review this fully soon.

Thanks for rebasing this so quickly

ariard commented:
> Only the last commit f057aed is the big scary change that removes cs_main locks all over the wallet, leaving us to hope that remaining locking is sufficient and that stretches of wallet code that used to run under cs_main aren't making assumptions about the tip not changing and other wallet threads not running.

Most of the code making assumptions about the tip is confined in the rescan logic, so I think this one should get the focus, you can grep for all methods fetching tip like `getHeight`, `getBlockHeight` `getX` and check no decision is made on return value of 2 different calls of them. Concerning other wallet threads, it should be cover by wallet lock in itself.

meshcollider commented:
Concept ACK

ryanofsky commented on src/wallet/rpcdump.cpp:
```diff
@@ -565,7 +565,7 @@ UniValue importwallet(const JSONRPCRequest& request)
             throw JSONRPCError(RPC_INVALID_PARAMETER, "Cannot open wallet dump file");
         }
-        Optional<int> tip_height = locked_chain->getHeight();
+        Optional<int> tip_height = pwallet->chain().getHeight();
```
Why use chain height instead of wallet height here and other places? Anywhere the wallet is locked and the chain isn't locked it would seem safer to use the wallet height.

ariard replied:
I think I can switch for the one in `CreateTransaction`, all left are tied to the rescan logic which doesn't call `BlockConnected` and so doesn't update `m_last_block_processed` accurately. I can rework a bit the rescan logic to make it work on m_last_block_processed_height but I felt it was a bit of scope and an increased burden for reviewers.

As you know I plan to rework the rescan logic after this PR. At term it would remove all `getHeight` callsites`.

ryanofsky replied:
Thanks, chain height seems fine for rescan logic, and is what I'd expect there. I think the other places should use wallet height if there isn't an explicit reason not to.

ariard replied:
Corrected on 24f40fc, there was only one occurrence in `CreateTransaction`. All others are tied to rescan logic, including ones in `rpcdump`/`rpcwallet`.

jkczyz commented:
When compiling with these changes, I get a new compilation warning related to holding `cs_wallet`:

```
  CXX      bitcoin_wallet-bitcoin-wallet.o
wallet/wallet.cpp:2528:61: warning: reading variable 'm_last_block_processed_height' requires holding mutex
      'cs_wallet' [-Wthread-safety-analysis]
    txNew.nLockTime = GetLocktimeForNewTransaction(chain(), m_last_block_processed_height);
                                                            ^
```

jonatack commented:
Approach ACK, light tested ACK as I have only made a cursory code review. It looks like you got all the call sites. Can `LockAssertion` in sync.h be removed? It appears to be unused after these changes.

Built with no warnings, ran tests and bitcoind with --enable-debug for the `-DDEBUG_LOCKORDER` compiler flag. No debug log warnings so far apart from the continual `n of last 100 blocks have unexpected version`. Would like to do the same rebased to current master but the rebase isn't clean; will wait for rebase.

```
2019-12-04T15:41:23Z Bitcoin Core version v0.19.99.0-24f40fc1af (debug build)
```

jonatack commented:
@jkczyz, what platform and configure flags did you use so I can try to reproduce? My build (Debian, --enable-debug, --enable-bench) didn't show that warning.

jonatack commented:
@jkczyz, what platform, compiler, and configure flags did you use so I can try to reproduce? My build (Debian, gcc 8.3, --enable-debug, --enable-bench) didn't show that warning. Perhaps I should use clang -c -Wthread-safety.

jonatack reviewed (commented):
Approach ACK, light tested ACK as I have only made a cursory code review. It looks like you got all the call sites. Can LockAssertion in `sync.h` be removed? It appears to be unused after these changes.

Built with no warnings, ran tests and bitcoind with --enable-debug for the `-DDEBUG_LOCKORDER` compiler flag. No debug log warnings so far apart from the continual `n of last 100 blocks have unexpected version` ones. Would like to do the same rebased to current master but the rebase isn't clean; will wait for rebase.

```
2019-12-04T15:41:23Z Bitcoin Core version v0.19.99.0-24f40fc1af (debug build)
```

jkczyz commented:
> @jkczyz, what platform, compiler, and configure flags did you use so I can try to reproduce? My build (Debian, gcc 8.3, --enable-debug, --enable-bench) didn't show that warning. Perhaps I should use clang -c -Wthread-safety.

@jonatack Built on MacOS using clang-1001.0.46.4. No special configuration flags. I performed `make clean && make` at a7aec7ad9 followed by `make` at 24f40fc1a to confirm the warning only appeared at the latter commit. I haven't tried building on my Linux box yet.

jkczyz commented:
> > @jkczyz, what platform, compiler, and configure flags did you use so I can try to reproduce? My build (Debian, gcc 8.3, --enable-debug, --enable-bench) didn't show that warning. Perhaps I should use clang -c -Wthread-safety.
> 
> @jonatack Built on MacOS using clang-1001.0.46.4. No special configuration flags. I performed `make clean && make` at [a7aec7a](https://github.com/bitcoin/bitcoin/commit/a7aec7ad97949a82f870c033d8fd8b65d772eacb) followed by `make` at [24f40fc](https://github.com/bitcoin/bitcoin/commit/24f40fc1afa9d25e5032ba4f8d1c0161e81dce3f) to confirm the warning only appeared at the latter commit. I haven't tried building on my Linux box yet.

Confirmed no warning on Ubuntu:

```
$ gcc --version
gcc (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
```

ariard commented:
Thanks for reviews,

@jkczyz it should be good now on df508be, following Russ remark I switchted `GetLocktimeForNewTransaction` to use `m_last_block_processed_height` but without remembering the reason I avoid to do it at first, which was to avoid tweaking lock order issue. I just changed `GetLocktimeForNewTransaction` after lock taking in `CreateTransaction`, it shouldn't change anything.

>  Can LockAssertion in sync.h be removed? 

@jonatack I would prefer to let it like this. We may need again this thread assertion helper in the future. IIRC clang thread safety detection only works on Mac for now

jonatack commented:
> @jonatack Built on MacOS using clang-1001.0.46.4. No special configuration flags. I performed `make clean && make` at [a7aec7a](https://github.com/bitcoin/bitcoin/commit/a7aec7ad97949a82f870c033d8fd8b65d772eacb) followed by `make` at [24f40fc](https://github.com/bitcoin/bitcoin/commit/24f40fc1afa9d25e5032ba4f8d1c0161e81dce3f) to confirm the warning only appeared at the latter commit. I haven't tried building on my Linux box yet.

@jkczyz Thanks! Found this warning with Debian/clang version 7.0.1-8:
```
  CXX      libbitcoin_common_a-merkleblock.o
wallet/wallet.cpp:2528:61: warning: reading variable 'm_last_block_processed_height' requires holding mutex 'cs_wallet' [-Wthread-safety-analysis]
    txNew.nLockTime = GetLocktimeForNewTransaction(chain(), m_last_block_processed_height);
                                                            ^
1 warning generated.
```
```
  CC            = /usr/bin/ccache clang
  CFLAGS        = -g -O2
  CPPFLAGS      =  -DDEBUG -DDEBUG_LOCKORDER  -Qunused-arguments
                   -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS
  CXX           = /usr/bin/ccache clang++ -std=c++11
  CXXFLAGS      =  -O0 -g3 -ftrapv  -Wstack-protector -fstack-protector-all  -Wall -Wextra
                   -Wformat -Wvla -Wswitch -Wformat-security -Wthread-safety-analysis
                   -Wrange-loop-analysis -Wredundant-decls  -Wno-unused-parameter
                   -Wno-self-assign -Wno-unused-local-typedef -Wno-deprecated-register
                   -Wno-implicit-fallthrough
```

jonatack commented:
> it should be good now on [df508be](https://github.com/bitcoin/bitcoin/commit/df508be1368e72a77b9ad50280a5a1c5d2e983b1)
> 
> IIRC clang thread safety detection only works on Mac for now

Yes, rebuilt and the warning is now gone in df508be. As I commented just above, I did reproduce it on 24f40fc with Debian and Clang (passing --enable-debug). Now testing rebased over current master.

practicalswift commented:
Would be interesting to see the impact (if any) on IBD :)

ariard commented:
@jonatack that's a compiler warning and `--enable-debug` is about turning on our deadlock detection with conditional compilation they are not related.

jkczyz commented:
I'm getting the warning at df508be on MacOS using clang as before.

ariard commented:
@jkczyz  at 3773b41, add a `AssertLockHeld` in `CreateTransaction` after taking `cs_wallet`, it should avoid clang false positive warning according to documentation "utility class for indicating to compiler thread analysis that a mutex is locked (when it couldn't be

jkczyz commented:
> @jkczyz at [3773b41](https://github.com/bitcoin/bitcoin/commit/3773b4159c37db993b9bb7c3f5d7dd7a51a39b09), add a `AssertLockHeld` in `CreateTransaction` after taking `cs_wallet`, it should avoid clang false positive warning according to documentation "utility class for indicating to compiler thread analysis that a mutex is locked (when it couldn't be determined otherwise)"

Looks like I had a problem on my end. I don't get the warning with df508be after all. Sorry about the confusion!

jonatack commented:
> @jonatack that's a compiler warning and `--enable-debug` is about turning on our deadlock detection with conditional compilation they are not related.

Thanks. Agreed. Verified locally: warning present on 24f40fc with Debian+Clang without passing --enable-debug.
```
  CC            = /usr/bin/ccache clang
  CFLAGS        = -g -O2
  CPPFLAGS      =   -U_FORTIFY_SOURCE  -D_FORTIFY_SOURCE=2  -Qunused-arguments
                    -DHAVE_BUILD_INFO  -D__STDC_FORMAT_MACROS
  CXX           = /usr/bin/ccache clang++ -std=c++11
  CXXFLAGS      =   -Wstack-protector -fstack-protector-all  -Wall -Wextra -Wformat
                    -Wvla -Wswitch -Wformat-security -Wthread-safety-analysis
                    -Wrange-loop-analysis -Wredundant-decls  -Wno-unused-parameter
                    -Wno-self-assign -Wno-unused-local-typedef -Wno-deprecated-register
                    -Wno-implicit-fallthrough   -g -O2
```

jonatack commented:
@ariard what changed from df508be to df17e36?

ariard commented:
@jonatack Adding the `AssertLockHeld` to be sure to not raise false positive warning from the compiler.

jonatack commented:
@ariard 3773b4159c37db993b9bb7c3f5d7dd7a51a39b09 builds without warnings or errors. Where exactly did you add the `AssertLockHeld`? I don't see it in CreateTransaction, nor in the diff of the last commit, and I'm not getting `git range-diff` to help me. Help a dumb reviewer out :D

ariard commented:
@jonatack src/wallet/wallet.cpp L2553 on head commit. But you should really recompile latest git to get `git range-diff` it changes your life!

instagibbs commented:
could you update the OP? It's slightly outdated due to dependent PRs being merged

ariard commented:
OP updated, see also for more context https://bitcoincore.reviews/16426.html

promag commented on src/interfaces/chain.cpp:
```diff
@@ -344,6 +331,7 @@ class ChainImpl : public Chain
     }
     std::unique_ptr<Handler> handleNotifications(Notifications& notifications) override
     {
+        LOCK(::cs_main);
```
3773b4159c37db993b9bb7c3f5d7dd7a51a39b09

Is this going to change in a follow up? Maybe add a comment explaining this lock?

ariard replied:
Adding a comment only for this method would seem obscure compare to other methods for future reviewers after refactoring, that's why it's in commit message "Lock cs_main in handleNotifications as it was relaying on its caller to do  so before", or do you have a different opinion ?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/16426#discussion_r356338388

> Adding a comment only for this method would seem obscure compare to other methods for future reviewers after refactoring

I think this lock is a little different than the locks in the other Chain methods because it is not obvious why locking is needed for something that appears to just be a memory allocation. If this lock really is necessary, having a comment in the code would be helpful to me at least. Or if it is possible to drop the lock or move it closer to where it might be needed in the [`NotificationsHandlerImpl` constructor](https://github.com/bitcoin/bitcoin/blob/3773b4159c37db993b9bb7c3f5d7dd7a51a39b09/src/interfaces/chain.cpp#L44-L48)  that would be even better.

ariard replied:
Okay after thinking about this, by removing lock, we may miss block connections between end of rescan and registering with validation interface. That was the meaning of the comment I think, which is about consistency of chain viewed by the wallet. To avoid that I moved `handleNotifications` before rescan and after `cs_wallet` lock taking. That's way block connections notifications are going to be pending until rescan is done. We may have block processing duplicata but at least we should guarantee wallet consistency. 

Long-term solution is to unify both rescan and Chain::Notifications under same interface.

ariard commented on src/wallet/wallet.cpp:
```diff
@@ -3403,16 +3403,14 @@ void CWallet::GetKeyBirthTimes(interfaces::Chain::Lock& locked_chain, std::map<C
+            for (const auto& keyid : GetAffectedKeys(txout.scriptPubKey, *spk_man)) {
+                // ... and all their affected keys
+                std::map<CKeyID, int>::iterator rit = mapKeyFirstBlock.find(keyid);
+                if (rit != mapKeyFirstBlock.end() && wtx.m_confirm.block_height < rit->second)
```
Is this a style requirement? IMO don't find code easier to read doing so.

kallewoof replied:
Yes, it is a style requirement. Same line or add `{}`.

From https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-c
> If an `if` only has a single-statement `then`-clause, it can appear on the same line as the `if`, without braces. In every other case, braces are required, and the `then` and `else` clauses must appear correctly indented on a new line.

ariard commented on src/wallet/wallet.cpp:
```diff
@@ -2445,13 +2445,14 @@ bool CWallet::FundTransaction(CMutableTransaction& tx, CAmount& nFeeRet, int& nC
     }
     constexpr int64_t MAX_ANTI_FEE_SNIPING_TIP_AGE = 8 * 60 * 60; // in seconds
-    if (locked_chain.getBlockTime(*locked_chain.getHeight()) < (GetTime() - MAX_ANTI_FEE_SNIPING_TIP_AGE)) {
+    auto locked_chain = chain.lock();
```
Going to let it like this, one of both commits have to clean it anyway unless adding a temporary one line is really gross..

ariard commented:
Thanks for review @promag, corrected some nits on f87558b.

ariard commented:
@practicalswift after a run on bitcoinperf, there is no seen performance change on IBD. It still may be an improvement with a heavy workload on wallet RPCs (but harder to simulate).

ryanofsky commented on src/interfaces/chain.h:
```diff
@@ -56,74 +51,59 @@ class Chain
+    //! included in the current chain.
+    virtual Optional<int> getBlockHeight(const uint256& hash) = 0;
+
+    //! Get block hash. Height must be valid or this function will abort.
```
For getBlockHash, getBlockTime, and getBlockMedianTimePast, I don't think it safe anymore for these methods to abort when an invalid height is passed. Previously it was possible to guarantee that an invalid height would never be passed because `cs_main` would be held before calling these (and if an invalid height was passed it indicated a bug). But now that cs_main isn't held, there could be a race condition where `invalidateblock` is called, or there is a reorg to a shorter chain that would cause aborts now.

A possible fix would be to implement error handling for race conditions in the places where these methods are called, maybe changing these methods to throw an exception or return `Optional<>` values or success `bools` to indicate failure instead of aborting.

Another solution would be to drop these methods, and replace them with a more general method similar to [`findBlock`](https://github.com/bitcoin/bitcoin/blob/ddecb671f0a41eaa8e6198f7e55172c1256bd978/src/interfaces/chain.h#L128-L137) that is capable of returning multiple pieces of information atomically so there aren't new race conditions the wallet has to handle now because it isn't holding `cs_main`.

ariard replied:
You're right, even if you need bad-timed reorgs to trigger them. After thinking about this, I picked the option to return `Optional<>` because that's one giving us more control. Throwing exception don't fit everywhere where an absent block isn't fatal. Return boolean would mean to pass pointer to values which is a bit different than other interface methods (like `getHeight`). 

On dropping these methods, I think `getBlockTime` would be the easiest one if we can store block time in confirmation struct of transactions but we're going to be limited by not being able to serialize them. `getBlockMedianTimePast` and `getBlockHash` can go by refactoring rescan logic.

ariard replied:
I've been through function triplet callsites again and I think using hash instead of height wouldn't make code safer that much. 

`getBlockHash` is always called by rescan code which making the assumption of chain lock not being held.

`getBlockTime`/`getBlockMedianTimePast`, given than block time must always be higher than parent MTP and not to be more than 2h in the future by consensus rules and than wallet code is using time range to take decision it doesn't matter than height match with block hash A or block hash B.

ryanofsky commented on src/interfaces/wallet.cpp:
```diff
@@ -60,16 +60,16 @@ WalletTx MakeWalletTx(CWallet& wallet, const CWalletTx& wtx)
 {
     WalletTxStatus result;
-    result.block_height = locked_chain.getBlockHeight(wtx.m_confirm.hashBlock).get_value_or(std::numeric_limits<int>::max());
+    result.block_height = wtx.m_confirm.block_height;
```
Is behavior changing here? Does this need need to be changed to continue returning `std::numeric_limits<int>::max()` in the case of an abandoned or unconfirmed transaction, or does it not matter?

ariard replied:
I think it doesn't matter, and maybe even confusing to display `std::numeric_limit<int>::max()` as block height. An abandoned one will have `is_abandoned` sets to true. An unconfirmed one will have `depth_in_main_chain` sets to 0. It seems to me meaningful enough for end user but I can still do a ternary and if `hashBlock` is null set it to fallback value if you think so.

ryanofsky replied:
> I think it doesn't matter, and maybe even confusing to display `std::numeric_limit<int>::max()` as block height. An abandoned one will have `is_abandoned` sets to true. An unconfirmed one will have `depth_in_main_chain` sets to 0. It seems to me meaningful enough for end user but I can still do a ternary and if `hashBlock` is null set it to fallback value if you think so.

I think anything that doesn't break the GUI is fine. If GUI treats 0x7FFFFFFFF and 0xFFFFFFFFF the same, the current code is fine. If GUI treats them differently, that should just be addressed somehow. I'm just asking because I'm not sure if this is a real change or not.

ariard replied:
Back to `result.block_height = wtx.m_confirm.block_height > 0 ? wtx.m_confirm.block_height std::numeric_limits<int>::max();`

Should avoid to break qt code, without further digging into it.

ryanofsky commented on src/interfaces/wallet.cpp:
```diff
@@ -330,14 +316,13 @@ class WalletImpl : public Wallet
-            block_time = locked_chain->getBlockTime(*height);
+        num_blocks = m_wallet->GetLastBlockHeight();
+        if (num_blocks >= 0) {
+            block_time = m_wallet->chain().getBlockTime(num_blocks);
```
I think to avoid the race condition mentioned https://github.com/bitcoin/bitcoin/pull/16426#discussion_r357761094, where there is a block reorg between the `CWallet::GetLastBlockHeight` and `Chain::getBlockTime` calls, it would be safer to avoid triggering asserts by getting the time by hash rather than height. You could tweak `GetLastBlock` to optionally return the block hash:

```c++
int GetLastBlock(uint256* block_hash = nullptr);
```

for

```c++
uint256 block_hash;
num_blocks = m_wallet->GetLastBlock(&block_hash);
if (num_blocks < 0 || !m_wallet.chain().findBlock(block_hash, nullptr, &block_time)) {
    block_time = -1;
}
```

There are probably other places in this PR where similar changes would be appropriate.

ariard replied:
Changed it in cbd2c1a, covered with other reorg mitgations.

ryanofsky commented on src/wallet/rpcdump.cpp:
```diff
@@ -358,8 +355,7 @@ UniValue importprunedfunds(const JSONRPCRequest& request)
 
-    auto locked_chain = pwallet->chain().lock();
-    Optional<int> height = locked_chain->getBlockHeight(merkleBlock.header.GetHash());
+    Optional<int> height = pwallet->chain().getBlockHeight(merkleBlock.header.GetHash());
```
I think the `LOCK(pwallet->cs_wallet);` line below should move above this `getBlockHeight()` call. Otherwise there could be a race condition if there is a reorg and the transaction is no longer confirmed by the time AddToWallet is called below, but the wallet considers it confirmed anyway.

I'm not sure if it would be a good idea, but perhaps the `CWallet::chain()` method should be annotated `EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)` to catch other cases like this.

ariard replied:
Thanks, I've modified it, but didn't add `EXCLUSIVE_LOCKS_REQUIRED` to the `CWallet::chain()` as it would change lock assumptions for RPC calls like `bumpfee` or `fundtransaction`..

ryanofsky commented:
I was reading the review club discussion, and it sounded like there was some confusion about what motivated the PR. I think it would help to simplify the PR description. If I were describing this PR, I would write something like:

<dl><dd><p>This change is intended to make the bitcoin node and its rpc, network, and gui interfaces more responsive while the wallet is in use. Currently, because the node's <code>cs_main</code> mutex is always locked before the wallet's <code>cs_wallet</code> mutex (to prevent deadlocks), <code>cs_main</code> currently stays locked while the wallet does relatively slow things like creating and listing transactions.</p>

<p>Switching the lock order so <code>cs_main</code> is acquired after <code>cs_wallet</code> allows <code>cs_main</code> to be only locked intermittently while the wallet is doing slow operations, so the node is not blocked waiting for the wallet.</p></dd></dl>

re: https://github.com/bitcoin/bitcoin/pull/16426#issue-299541584

> `Chain::Lock` reason was mostly a temporary interface to help separation of Wallet from Node (#10973) without modifying behavior of legacy wallet code relying on chain state. It was a relief of review burden in the aforementioned separation PR, at the cost of keeping synchronicity between Wallet and Node execution.
> 
> Caching more state in CWallet (#15931), let us remove the more used calls like `getBlockDepth` and move toward a more asynchronous interface. The current PR 1) move all remaining Chain::Lock methods to simple one and ensure `cs_main` lock is taken inside the interface 2) remove locking of `cs_main` inside any wallet code, which is another asynchronicity gain.
> Lock order has to be swapped atomically at once to avoid potential deadlock issues raised by thread sync verification stuff.

Especially now that #15931 is merged, this mostly feels like historical information that doesn't help make the PR description more understandable. I would probably drop these paragraphs from the PR description or maybe move them to footnotes.

> Some future works which would enhance Chain interface:
> 
> * move `ScanForWalletTransactions` and most we can of rescanning logic on the node side,  would let us remove directly methods like `guessVerificationProgress` or `getBlockHeight`, meanwhile relieving of the the hardship of multiple simultaneous rescanning processing
> * split `listsinceblock` into a wallet part and node one, #15931 caching block height in tx, I think that's possible
> * move fee estimation queries to a notification interface and let the wallet cache it, fee estimation being driven by node mempool, wallet should be the consumer
> * obviously remove the RPC specific methods by letting the wallet processing RPC requests without going through the node, need to add `--server`  and `--rpcserver` options

I'm also not sure so much detail about future todos belongs in the PR description here. I've been trying to keep a list of cleanup todos at the top of src/interfaces/chain.h, so you could definitely merge this list into that list, and keep discussion of followup changes in the PR description more brief and basic.

> To review the present PR, most of getting right the move is ensuring any `LockAssertion` in `Chain::Lock` method is amended as a `LOCK(cs_main)`. And in final commit, check that any wallet code which was previously locking the chain is now calling a method, enforcing the lock taking job. So far the only exception I found is `handleNotifications`, which should be corrected.

This paragraph is helpful and seems useful to keep, except I'm not clear about the `handleNotifications` comment. Would that be addressed by following up to https://github.com/bitcoin/bitcoin/pull/16426#discussion_r355227118?

fjahr commented on src/wallet/rpcdump.cpp:
```diff
@@ -790,9 +780,9 @@ UniValue dumpwallet(const JSONRPCRequest& request)
-    file << strprintf("#   mined on %s\n", tip_height ? FormatISO8601DateTime(locked_chain->getBlockTime(*tip_height)) : "(missing block time)");
+    int tip_height = pwallet->GetLastBlockHeight();
+    file << strprintf("# * Best block at time of backup was %i (%s),\n", tip_height, tip_height >= 0 ? pwallet->chain().getBlockHash(tip_height).ToString() : "(missing block hash)");
+    file << strprintf("#   mined on %s\n", tip_height >= 0 ? FormatISO8601DateTime(pwallet->chain().getBlockTime(tip_height)) : "(missing block hash)");
```
nit: `tip_height` checks not needed here, see above

fjahr reviewed (commented):
ACK f87558b49bc63922a70c9ff02c33c0b1ed1f44b1

Reviewed code, built and ran tests locally, ran it as my testnet node and manually tested some wallet RPCs.

One minor nit (occuring multiple times), feel free to ignore.

ryanofsky commented:
A suggestion for other reviewers might be to review last commit first: "Remove locked_chain from CWallet, its RPCs and tests" (f87558b49bc63922a70c9ff02c33c0b1ed1f44b1), since this is the commit that actually changes behavior, while other commits are moving and refactoring code.

Note to Antoine: If it would make it easier to update the PR, I think it'd be fine to squash it into just one or two commits. Personally I'm mostly reviewing this PR by looking at the overall diff, because it's less work to focus on the end result when reviewing parts of the code that are changing more than once.

Commit:
[wallet] Remove locked_chain from CWallet, its RPCs and tests

This change is intended to make the bitcoin node and its rpc, network
and gui interfaces more responsive while the wallet is in use. Currently
because the node's cs_main mutex is always locked before the wallet's
cs_wallet mutex (to prevent deadlocks), cs_main currently stays locked
while the wallet does relatively slow things like creating and listing
transactions.

This commit only remmove chain lock tacking in wallet code, and invert
lock order from cs_main, cs_wallet to cs_wallet, cs_main.
must happen at once to avoid any deadlock. Previous commit were only
removing Chain::Lock methods to Chain interface and enforcing they
take cs_main.

Remove LockChain method from CWallet and Chain::Lock interface.

ariard commented:
Thanks one thousands times @ryanofsky for review and advices. Implemented most of them, including updating OP and tracking TODO in `interfaces/chain.h`. Also squash a bunch of commits, only non-squashed are the ones doing refactor. 

I'm still looking for low-hanging fruits to dry-up the interface like dropping `checkFinalTx` to ease review.

ariard commented on src/interfaces/chain.cpp:
```diff
@@ -234,13 +124,104 @@ class ChainImpl : public Chain
+    {
+        LOCK(::cs_main);
+        CBlockIndex* block = ::ChainActive()[height];
+        assert(block != nullptr);
```
Ooops, my bad, thanks should be corrected on latest tip.

ryanofsky replied:
In commit "[wallet] Move getBlockHash from Chain::Lock interface to simple Chain" (c6fa030ca8a282a7906260186e93d8c8902aa5a4)

re: https://github.com/bitcoin/bitcoin/pull/16426#discussion_r362999187

> Ooops, my bad, thanks should be corrected on latest tip.

Assert is still present (maybe added back again)

ariard replied:
Corrected, effectively that's a rebase error. It's still present on master but removing `locked_chain` means you may have a reorg while calling `getBlockHash` in wallet creation rescan.

ariard commented:
Updated e4da874. Picked up Russ's last review suggestions/findings, thanks.

kallewoof reviewed (commented):
Code review up to and including eff956c1d63975d386d68d0fe764b372a2ed81ea (Move getBlockHash from Chain::Lock interface to simple Chain).

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2914,9 +2888,8 @@ static UniValue listunspent(const JSONRPCRequest& request)
         cctl.m_min_depth = nMinDepth;
         cctl.m_max_depth = nMaxDepth;
-        auto locked_chain = pwallet->chain().lock();
         LOCK(pwallet->cs_wallet);
```
Move this up and remove the lock in L2922.

ariard replied:
You mean the main wallet lock just after ? I think some other wallet locks can also be removed in other places and that would be better to do it in a follow-up PR with dedicated thinking.

ariard commented:
Thanks @kallewoof for review, took most of the suggestions on ac05c8b, some of them were already fixed in latter commits anyway, I can squash commits more if it suits more (not my feeling but if it fits reviewers better let's go for it)

jnewbery reviewed (commented):
I've started skimming through the changes here. I have a few trivial comments so far.

ryanofsky reviewed (commented):
Thanks for the updates Antoine, and sorry I've been taking so long to review this. I've mostly just been stuck on the point about block heights not being a reliable way to refer to blocks when the chain isn't locked. But I did some work this week to try to figure things out.

re: https://github.com/bitcoin/bitcoin/pull/16426#discussion_r363000179
> I've been through function triplet callsites again and I think using hash instead of height wouldn't make code safer that much.

I think I agree that safety isn't a big issue most of these cases because failures should be obvious and just retrying anything that fails should fix things. But I guess I look at this type of safety as more of a binary choice than a sliding scale. I.e. code can either be correct and have race conditions, or be incorrect and not have them.

So I've been working on a change built on top of this PR and #15719 to get rid of all the Chain height methods: 0bc4a8d9530804f82fc6aa272c3cb543ed9f2263. Working on this has helped me understand this PR better, and I think I can break some pieces out of that commit to smaller PRs that remove `Chain::Lock` usages, so this PR can be smaller, safer, and easier to understand. Even though this will have the drawback of making overall changes larger, I think the end result should be safer, and individual changes easier to verify.

Just wanted to give this update. I should have more comments and maybe some separate PRs related to this one soon.

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/16426#pullrequestreview-343566228

> I think I can break some pieces out of that commit to smaller PRs that remove `Chain::Lock` usages

I started doing this in draft PR #17954. Will continue to update

ryanofsky reviewed (approved):
Code review semi-ACK ac05c8b8119e4f0a055a26a98d69e4b6cbfdd2c1

I finally reviewed this whole PR and would be ok with merging it in its current form. But my preference would be to rebase this PR on top of #17954 and merge it after. Reasons:

1. This PR seems to add a lot of little race conditions, looking up blocks by height instead of hash when the chain state is not locked. I don't think any of these are likely to ever trigger in practice or cause serious problems if they did trigger, but it leaves the code in an awkward state to have these.

2. Ordering #17954 before this PR would reduce churn in Chain interface methods. It would avoid adding temporary methods like getBlockTime that we are just going to delete later. (For the same reason it also might be nice to merge #17443 before this, to avoid needing to add `Chain::checkFinalTx` which that PR would get rid of.)

3. Ordering #17954 before this PR would reduce churn in wallet code and make the changes there easier to review. #17954 is structured to update complicated functions like `listsinceblock` and `CWallet::ScanForWalletTransactions` once in individual commits. By contrast this PR updates the same functions multiple times in different commits, which I think can make it hard to see how the changes fit together. If #17954 is ordered first, most of these changes should be unnecessary.

laanwj commented:
This needs rebase and addressing @ryanofsky's comments.

ryanofsky commented:
> This needs rebase and addressing @ryanofsky's comments.

This is basically just blocked on #17954.

Antoine has a rebased version of this PR on top of #17954 and #17443, which both simplify this PR, in a working branch. https://github.com/bitcoin/bitcoin/compare/master...ariard:2020-02-wallet-standalone

The relevant commits in that branch make a smaller version of this PR:

- 23c478e0515571ef8291a337d9c351d487b059e3 Move getBlockHeight from Chain::Lock interface to simple Chain
- 7a17b679aad171b67426d74b445cec3ac16c0043 Move getBlockHash from Chain::Lock interface to simple Chain
- 08c90335510c95f845c840407058699385abf93c Move multiple methods from Chain::Lock interface to simple Chain
- d2fa1e1ad3bad97436142262964707ab87f395f1 Remove locked_chain from CWallet, its RPCs and tests

(Later commits on that branch have to do with implementing rescan logic outside of the wallet so it is cleaner and can allow txindex, blockfilterindex, and multiple wallet rescans to all happen in parallel without each wallet and index having to loop and reread the same blocks. The nonwallet parts are in another branch: https://github.com/bitcoin/bitcoin/compare/master...ariard:2019-08-rescan-index-refactor)

ariard commented:
@laanwj Thanks for the push-up, we agreed with Russ to get #17954 first, almost of all of his comments are related in way #17954 do it better if rebased on top. Also it would let me remove few commits. Updated PR description.

ryanofsky commented:
@ariard, it would be great to rebase this now that #17954 is merged. It should be simpler now, but good to merge early in the release cycle since it changes wallet locking in a big way

ariard commented:
@ryanofsky Thanks for the push-up after #17954, it simplifies this PR so much. Let me know if you think I should split/aggregate commits.

Note, I remove `findPruned` and `findFork` as they were useless after your PR.

MarcoFalke commented:
https://travis-ci.org/github/bitcoin/bitcoin/jobs/676012190#L7078

```
2020-04-17T04:31:03.393017Z POTENTIAL DEADLOCK DETECTED

2020-04-17T04:31:03.393067Z Previous lock order was:

2020-04-17T04:31:03.393091Z  (1) cs_wallet wallet/wallet.cpp:1667 (in thread )

2020-04-17T04:31:03.393138Z  (2) cs_main interfaces/chain.cpp:218 (in thread )

2020-04-17T04:31:03.393211Z Current lock order is:

2020-04-17T04:31:03.393233Z  (2) ::cs_main wallet/test/wallet_tests.cpp:464 (in thread )

2020-04-17T04:31:03.393275Z  (1) wallet->cs_wallet wallet/test/wallet_tests.cpp:464 (in thread )

Assertion failed: detected inconsistent lock order at sync.cpp:124, details in debug log.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -3998,9 +3973,18 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(interfaces::Chain& chain,
+    // but we guarantee at least than wallet state is correct after notifications delivery.
+    // This is temporary until rescan and notifications delivery are unified under same
+    // interface.
+    walletInstance->m_chain_notifications_handler = walletInstance->chain().handleNotifications(walletInstance);
```
In commit "[wallet] Remove locked_chain from CWallet, its RPCs and tests" (ef8c6ca60767cac589d98ca57ee33179608ccda8)

Note I wrote a test to detect the race condition this early `handleNotifications` call prevents: 25651aad58b1f6e543f1ad565d821de46268e724 ([branch](https://github.com/ryanofsky/bitcoin/commits/review.16426.5-edit)). Most of the test works on master so I made a separate PR #18727 for it. There's just a few lines in 25651aad58b1f6e543f1ad565d821de46268e724 not in that PR, helping the test work with this PR

ryanofsky commented on src/interfaces/chain.h:
```diff
@@ -135,6 +132,9 @@ class Chain
     virtual Optional<int> getBlockHeight(const uint256& hash) = 0;
 
+    //! Get block hash. Height must be valid or this function will abort.
+    virtual uint256 getBlockHash(int height) = 0;
```
In commit "[wallet] Move getBlockHash from Chain::Lock interface to simple Chain" (c6fa030ca8a282a7906260186e93d8c8902aa5a4)

Note: After this commit, the only remaining Chain::getBlockHash() calls are in CreateWalletFromFile and are removed in #15719. I expect this PR will get merged before #15719, so just making a note about this.

ariard replied:
Yes I have also #17484, #17443 which remove more interface methods to skim in after.

ryanofsky commented on src/interfaces/chain.cpp:
```diff
@@ -234,6 +225,15 @@ class ChainImpl : public Chain
     }
+    Optional<int> getHeight() override
+    {
+        WAIT_LOCK(cs_main, lock);
```
In commit "[wallet] Move getHeight from Chain::Lock interface to simple Chain" (b4eb6aee0d259c3586894694a67928759503c823)

This is fine, but in this commit and other commits throughout this PR could simplify `WAIT_LOCK(cs_main, lock)` to just `LOCK(cs_main)` since `lock` object is unused

ariard replied:
Right, only difference between them is what is logged for debug, attributing a name for the lock for `WAIT_LOCK(cs_main, lock)`? They both declare `DebugLock<decltype>`. Updating

ryanofsky commented on src/interfaces/chain.h:
```diff
@@ -135,6 +105,28 @@ class Chain
 
+    //! Check that the block is available on disk (i.e. has not been
+    //! pruned), and contains transactions.
+    virtual bool haveBlockOnDisk(int height) = 0;
```
In commit "[wallet] Move methods from Chain::Lock interface to simple Chain" (22d0fb127e99d6cdcaa003785a7278f58038b465)

Note: After this commit, the only remaining Chain::haveBlockOnDisk, findFirstBlockWithTimeAndHeight, getTipLocator, and findLocatorFork calls are in CreateWalletFromFile and are removed in #15719. I expect this PR will get merged before #15719, so just making a note about this.

Similarly, the Chain::checkFinalTx method can go away after #17443, which removes all those calls

ryanofsky reviewed (approved):
Code review ACK ef8c6ca60767cac589d98ca57ee33179608ccda8. Looks great! This PR now just has safer changes and I think is easier to review. I left a bunch of notes and maybe two minor suggestions, which you should feel free to ignore. This does need rebase but otherwise looks good to merge

ariard commented on src/interfaces/chain.h:
```diff
@@ -135,6 +130,11 @@ class Chain
+    //! Get current chain height, not including genesis block (returns 0 if
+    //! chain only contains genesis block, nullopt if chain does not contain
+    //! any blocks)
+    virtual Optional<int> getHeight() = 0;
```
Okay looking through #15719, this solve the issue by doing wallet rescan on the server-side so you don't need to get current tip.

ariard commented on src/interfaces/wallet.cpp:
```diff
@@ -335,7 +335,7 @@ class WalletImpl : public Wallet
         auto mi = m_wallet->mapWallet.find(txid);
         if (mi != m_wallet->mapWallet.end()) {
-            num_blocks = locked_chain->getHeight().get_value_or(-1);
+            num_blocks = m_wallet->GetLastBlockHeight();
```
`m_last_block_processed_height` is already initialized at `-1`, so behavior change is in case wallet isn't at tip right ? If yes, it should be more accurate.

ariard commented on src/interfaces/chain.h:
```diff
@@ -135,6 +130,11 @@ class Chain
+    //! Get block height above genesis block. Returns 0 for genesis block,
+    //! 1 for following block, and so on. Returns nullopt for a block not
+    //! included in the current chain.
+    virtual Optional<int> getBlockHeight(const uint256& hash) = 0;
```
Okay going to let it as it is, given it would need an extension of FoundBlock, and this being gauged by other reviewers. Let's keep diff straight.

ariard commented:
Rebased at d6d6632, with some Russ suggestions. Test failure was due to deadlock in test framework introduced by a new `LOCK2` and not cached at last rebae.

MarcoFalke commented:
Tests are still failing

promag reviewed (commented):
Code review ACK 0716c9cb68de74cd6e3f791dbe2aec103f70e292. Still running some tests. 

So loading an old wallet (that needs rescan) still freezes the node until the wallet catches up. Maybe it could try to register as late as possible? Like before the last rescanned block?

MarcoFalke commented:
needs rebase

fjahr commented:
re-ACK 0716c9cb68de74cd6e3f791dbe2aec103f70e292

Did another full review of the code and built and ran tests locally.

kallewoof commented:
Typo in commit 4509c05eab82dda405b4265838a61ed7ca302bf7 message: `chain acces` → `chain access`.

kallewoof commented on src/interfaces/chain.h:
```diff
@@ -51,79 +46,74 @@ class Wallet;
+//!   wallet cache it, fee estimation being driven by node mempool, wallet
+//!   should be the consumer.
+//!
+//! * The `guesVerificationProgress`, `getBlockHeight`, `getBlockHash`, etc
```
This typo is still present in current code.

kallewoof reviewed (commented):
Code review ACK 0716c9cb68de74cd6e3f791dbe2aec103f70e292

ariard commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -240,6 +240,10 @@ BOOST_FIXTURE_TEST_CASE(importwallet_rescan, TestChain100Setup)
         std::shared_ptr<CWallet> wallet = std::make_shared<CWallet>(chain.get(), WalletLocation(), WalletDatabase::CreateDummy());
         LOCK(wallet->cs_wallet);
         wallet->SetupLegacyScriptPubKeyMan();
+        {
```
Yeah was a rebase error due to lock being added in between in another PR. Fixed, thanks!

ariard commented:
Rebased 49ecec4. Thanks a lot @promag @fjahr @kallewoof for reviews! If you can re-ACK there should be no change apart of dropping locks added in `importdescripotrs` in the meanwhile.

> So loading an old wallet (that needs rescan) still freezes the node until the wallet catches up. Maybe it could try to register as late as possible? Like before the last rescanned block?

It's planned in future works, see https://github.com/bitcoin/bitcoin/pull/15719 by Russ. I've also a branch doing it in an alternative way.

ryanofsky reviewed (approved):
Code review ACK 49ecec42ea81e9cbf3970416ead79e18863dea87. Changes since last review: rebase, switching from WAIT_LOCK to LOCK, dropping assert in getBlockHash (still segfaults but this is temporary https://github.com/bitcoin/bitcoin/pull/16426#discussion_r412494509), switched to consistent lock order in ListCoins test

This has a few recent ACKs so I think may be ready to merge or very close

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/16426#pullrequestreview-400824121

> So loading an old wallet (that needs rescan) still freezes the node until the wallet catches up.

@promag is this comment saying the PR is introducing a new freeze which is not in master, or is this talking about an existing freeze? Either way I think #15719 may address it, but it would be good to know what the exact issue is (steps to reproduce and observed behavior)

jonatack commented:
Finishing up review of Signet PR, I'll try to review this tomorrow @ariard

MarcoFalke commented on src/interfaces/chain.cpp:
```diff
@@ -234,6 +227,12 @@ class ChainImpl : public Chain
+    {
+        LOCK(::cs_main);
+        CBlockIndex* block = ::ChainActive()[height];
+        return block->GetBlockHash();
```
why remove the assert and make it undefined behavior?

27b1af965e

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/16426#discussion_r417366808

> why remove the assert and make it undefined behavior?
> 
> [27b1af9](https://github.com/bitcoin/bitcoin/commit/27b1af965ed9272496787663ea2869f05c2e5342)

I think it came from a suggestion I made https://github.com/bitcoin/bitcoin/pull/16426#discussion_r412493917 that wasn't clear  about a desire not to crash :smile:. But this is only called one place where it's extremely unlikely to trigger (there'd have to be a very specific thread pre-emption at the same time as a reorg at the same time as loading a wallet)

https://github.com/ariard/bitcoin/blob/49ecec42ea81e9cbf3970416ead79e18863dea87/src/wallet/wallet.cpp#L3927-L3929

Also this method should be removed with #15719

MarcoFalke replied:
> But this is only called one place where it's extremely unlikely to trigger

Sounds like an excellent use-case for `assert` instead of extremely unlikely undefined behavior. Especially given that the assert was there in the fist place.

I don't see why removing the assert can simply wait until #15719

jonatack replied:
Agree with @MarcoFalke here.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/16426#discussion_r417555069

> I don't see why removing the assert can simply wait until #15719

Of course. My suggestion was never to remove the assert by segfaulting instead. Originally the suggestion was to return Optional https://github.com/bitcoin/bitcoin/pull/16426#discussion_r357761094. Just pointing out what's done here isn't too important because it's basically impossible to hit this condition and there's already a PR deleting the method

ariard replied:
By removing `cs_main` lock taking you may have now a block nullptr in the case where between `getHeight`/`findFirstBlockWithTimeAndHeight` and `getBlockHash` calls you have a thread taking `cs_main` lock, a reorg happening while at same time wallet loading. It's really unlikely but glad #15719 is cleaning this as a follow-up.

MarcoFalke commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -586,6 +586,11 @@ BOOST_FIXTURE_TEST_CASE(wallet_disableprivkeys, TestChain100Setup)
+        auto locked_chain = chain->lock();
+        LOCK(wallet->cs_wallet);
+        wallet->SetLastBlockProcessed(::ChainActive().Height(), ::ChainActive().Tip()->GetBlockHash());
+    }
```
Why is this needed? 7c73c4888f

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/16426#discussion_r417374233

> Why is this needed? [7c73c48](https://github.com/bitcoin/bitcoin/commit/7c73c4888f5c4c3f24a180f2841429318463db7d)

It does seem not to be needed anymore. Test passes for me with it commented out.

MarcoFalke commented on src/interfaces/wallet.cpp:
```diff
@@ -60,16 +60,16 @@ WalletTx MakeWalletTx(CWallet& wallet, const CWalletTx& wtx)
     result.lock_time = wtx.tx->nLockTime;
-    result.is_final = locked_chain.checkFinalTx(*wtx.tx);
-    result.is_trusted = wtx.IsTrusted(locked_chain);
+    result.is_final = wallet.chain().checkFinalTx(*wtx.tx);
```
what is the risk of this getting out-of-sync with `result.block_height` or other members? 9229faa752

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/16426#discussion_r417376070

> what is the risk of this getting out-of-sync with `result.block_height` or other members like `result.is_trusted`? [9229faa](https://github.com/bitcoin/bitcoin/commit/9229faa752cdd7bc778a67a638effb273ad7de19)

I think only risk is gui briefly displaying transaction as final in the transaction HTML description before wallet processing catches up and other fields in the HTML description reflect that. This is probably a minor concern but #17443 should fix it

MarcoFalke replied:
Oh, I wasn't aware this is only used for the gui

MarcoFalke commented:
Approach ACK 49ecec42ea 🛒

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

Approach ACK 49ecec42ea 🛒
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgcNgwAp3x7yEnzet/P74Z6x4BR0H3RIBb2n+oW1iOR8i1kkCEMviMketbAha8S
vziSxJJ45p7uQBEvetTRb3jlErpB6/tNkp7PAEVep204RTpQX03IlIrfVNg6s6Am
I5YF/m3DYJ2HQ5lyiTwk0+mQ5wgOKPYszYHRrxcV+x/qfxTwNN9WZnRp6TbEpiuV
s3B87BMz8olNUCjawcG4Y0jVahetsgH90e2wFUD0I1ERWfktHSNQ9CQaBOVWFU9L
nY1KtYxKE0KHPnctojN1tOeltcUUa5jQB+ezks6FACOKOs53yzG28QXzviiWD5gx
EThUTcn6DB/H4i7Z3QriaYsF7DCECvO+WDaHBZY+kMoCbpP6c3kCkPPD9axv9kZY
kA+gyIlzTqdJ7H2w2s/wsGffjiviYaowXd2R4bEuIqTlAdzt6xy+7hjMCbb5kjou
fB1r8eg3o6O2IM1mbWkZdH0tNK/itkxitMdvOICDk6bPAIk1zlEw6jQKa51hr61l
aUnAG1/b
=RG4E
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `9b0f6482a573980d5bcd1e554e17ef16e67861b355c26c79f7accca02eb8219b  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401089b0f6482a573980d5bcd1e554e17ef16e67861b355c26c79f7accca02eb8219bf01074813f231661294df7dcb66f2cc0354908fff01043a0a03a7b18c02dde2c90d64a64432e08f1045ea995b3f008e61eb2e4634c30430083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010a40f9d28f4d65285d396111c9d5a98f708f1045ea995b3f00844c2cc1bcbfa06050083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010d035d812b3ce46463f6dfa4334d1cfee08f1045ea995b3f008cd71f4c5ab8d33c60083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df010e7578ce2ae28abe11bb022cbcfce079c08f1045ea995b6f008b7578b9f92b373450083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

fjahr commented:
Re-ACK 49ecec42ea81e9cbf3970416ead79e18863dea87 but agree with https://github.com/bitcoin/bitcoin/pull/16426#discussion_r417374233 and https://github.com/bitcoin/bitcoin/pull/16426#discussion_r417555069 as worthwhile updates before merging.

jonatack commented:
Handy to have #18727 merged in. Concept ACK, agree with @fjahr above, will re-look tomorrow as it's getting late.

MarcoFalke commented:
Tests fail. Needs rebase

kallewoof commented:
Re-ACK based on the diff below (derived by rebasing the old commit on master and the git diffing this against the resulting state):

```diff
diff --git a/src/interfaces/chain.h b/src/interfaces/chain.h
index ec256d7b2..e33fe54ac 100644
--- a/src/interfaces/chain.h
+++ b/src/interfaces/chain.h
@@ -72,7 +72,7 @@ public:
 //!   wallet cache it, fee estimation being driven by node mempool, wallet
 //!   should be the consumer.
 //!
-//! * The `guesVerificationProgress`, `getBlockHeight`, `getBlockHash`, etc
+//! * The `guessVerificationProgress`, `getBlockHeight`, `getBlockHash`, etc
 //!   methods can go away if rescan logic is moved on the node side, and wallet
 //!   only register rescan request.
 class Chain
diff --git a/src/wallet/rpcdump.cpp b/src/wallet/rpcdump.cpp
index 7e46cb0c9..c863d2253 100644
--- a/src/wallet/rpcdump.cpp
+++ b/src/wallet/rpcdump.cpp
@@ -1664,7 +1664,6 @@ UniValue importdescriptors(const JSONRPCRequest& main_request) {
     bool rescan = false;
     UniValue response(UniValue::VARR);
     {
-        auto locked_chain = pwallet->chain().lock();
         LOCK(pwallet->cs_wallet);
         EnsureWalletIsUnlocked(pwallet);

@@ -1693,7 +1692,6 @@ UniValue importdescriptors(const JSONRPCRequest& main_request) {
     if (rescan) {
         int64_t scanned_time = pwallet->RescanFromTime(lowest_timestamp, reserver, true /* update */);
         {
-            auto locked_chain = pwallet->chain().lock();
             LOCK(pwallet->cs_wallet);
             pwallet->ReacceptWalletTransactions();
         }
```

ryanofsky commented:
Note: after #18727 the last commit here will cause https://github.com/bitcoin/bitcoin/blob/a66ba6d029b3948dec0cf092bb844f70c16b07dc/src/wallet/test/wallet_tests.cpp#L729 to fail.

But I previously posted a version of that test which works with this PR in 25651aad58b1f6e543f1ad565d821de46268e724 ([branch](https://github.com/ryanofsky/bitcoin/commits/review.16426.5-edit)) in https://github.com/bitcoin/bitcoin/pull/16426#discussion_r412461100, so there should be no new code to write if you take the test code from that commit:

https://github.com/ryanofsky/bitcoin/blob/25651aad58b1f6e543f1ad565d821de46268e724/src/wallet/test/wallet_tests.cpp#L669-L771

ryanofsky commented:
Rebased 49ecec42ea81e9cbf3970416ead79e18863dea87 -> c524664df16057763eaf2e5805af9bf8975ac3f8 ([`pr/rev.1`](https://github.com/ryanofsky/bitcoin/commits/pr/rev.1) -> [`pr/rev.2`](https://github.com/ryanofsky/bitcoin/commits/pr/rev.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/rev.1-rebase..pr/rev.2)). No conflicts and no changes other than adding back getBlockHash assert to resolve https://github.com/bitcoin/bitcoin/pull/16426#discussion_r417366808 and adding test code which was previously posted https://github.com/bitcoin/bitcoin/pull/16426#discussion_r412461100 so the new test from #18727 passes

MarcoFalke commented:
Thanks @ryanofsky

ariard commented:
Whaao thanks @ryanofsky for rebase! Took your branch at 6a72f26 while fixing @MarcoFalke 's https://github.com/bitcoin/bitcoin/pull/16426#discussion_r417374233.

ryanofsky reviewed (approved):
Code review ACK 6a72f26968cf931c985d8d4797b6264274cabd06. Only difference compared to the rebase I posted is reverting unneeded SetLastBlockProcessed change in wallet_disableprivkeys test

MarcoFalke commented:
re-ACK 6a72f26968 🔏

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK 6a72f26968 🔏
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhtVAwAprBpnjBnEpN+ACT69Dfo+rp5Gz1Ew6sRxvJr88Zkc/GSfZ9rCTlcrKIH
TXRD5CR4ba8Cj9zQSF4wo/LmlqwqJlgEZnhil5qRi+Cpo/nqJEoJaajmFQBjDOJW
7E68hxIxb9r9YF1G81gx71dLSHwUgx8RZp3POdZXInUigXk+XJJRwIlUckheMgXH
vF+als8TE+Fwv2gNyu03ggEAbNtFBi20yB/xLCApBcI6Ec3HDuj76JjhnqYQEetF
nxowDiw4EfUqwyD/XUek/OKQaXelynRoMhC4Xr+n1DNaWePoNune3qD/AGi2IF0v
VYleEGI01uPSryujDjo+o6HwWliKv7ooYjqSq4MAO387rE+hpVh368hfLqTwaPps
AeH12GMpgeALphIsqKzhd62PdeDUHxApvglj1zUQsuhSukx5ojX8oGNokWfA0slj
E+QKUT67kbM+2/s4kIJC3PLeMd97iYZfSanWrKTByjoecn42xf6hn15dldSsiqGM
KV4l6ZuH
=FnnV
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `e9607093f00416bfc4093550ccd6a4d573cc579bde37edf8d5b8a119a0b88ea9  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108e9607093f00416bfc4093550ccd6a4d573cc579bde37edf8d5b8a119a0b88ea9f010cc7087f2fed326092658ba67a0a8923708fff010478a4de1c9c74ad2fa6cb52f209c0d2e08f1045eab547cf008b35f11db7aa706260083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01048e3ca588fd3bc1060ce1e166511522c08f1045eab547cf008371ab94985f2aeb20083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01048e5ac49237812f2fab148658f104d0b08f1045eab547cf008781d4b0444b47c9c0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df01053c8d6d5619c7e708373e0243f3b34a508f1045eab547cf008ccdd970ed5951fa40083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

fjahr commented:
re-ACK 6a72f26968cf931c985d8d4797b6264274cabd06

Reviewed latest changes, built and ran tests locally.

ariard commented:
There is a stale test on some disk I/O issue, other are valid. If you can relaunch.

jonatack commented:
Congrats on the merge @ariard! Sorry, I was still reviewing the BIP324 PR.

MarcoFalke commented:
Review still welcome :) 

This had three explicit ACKs, one implicit re-ACK and a bunch of Concept ACKs. So beside being ready for merge, and also in light of the rebase hell this went through in the past, with the risk of (silent) merge conflicts putting it back in there, I went ahead and merged it.

ryanofsky commented:
> Review still welcome :)

Testing is useful too because all the wallet code that was originally written assuming cs_main was locked and state couldn't change in the background is now in an a more asynchronous world.

I'm also still curious about @promag's comment in https://github.com/bitcoin/bitcoin/pull/16426#pullrequestreview-400824121 and possible effects on the GUI. (Unclear if the comment is saying there's a new GUI freeze.)

jonatack commented:
All good, I agree!

promag commented:
> I'm also still curious about @promag's comment in #16426 (review) and possible effects on the GUI.

@ryanofsky missed this, AFAICT this doesn't add new issues to the GUI.

hebasto commented:
FWIW, this PR causes the issue #19049.
